File: clients/android/app/src/main/java/live/yunmao/app/util/GrayHit.kt
package live.yunmao.app.util

// 与后端 pkg/yunmao/featureflags Hash100 / iOS GrayHit 完全等价的 FNV-1a 32bit。
object GrayHit {
    fun hash100(key: String): Int {
        var hash: Int = 0x811C9DC5.toInt()
        for (b in key.toByteArray(Charsets.UTF_8)) {
            hash = hash xor (b.toInt() and 0xff)
            hash = (hash * 0x01000193).toInt()
        }
        // 取无符号 mod 100
        val unsigned = hash.toLong() and 0xFFFFFFFFL
        return (unsigned % 100).toInt()
    }

    fun inGrayPercent(roomId: String, percent: Int): Boolean {
        if (percent <= 0) return false
        if (percent >= 100) return true
        return hash100(roomId) < percent
    }
}

File: clients/ios/YunmaoApp/Sources/YunmaoApp/Util/GrayHit.swift
// 与 pkg/yunmao/featureflags FNV1a Hash100 等价的 client-side 灰度命中。

import Foundation

public enum GrayHit {
    /// 与后端 Hash100 完全等价：FNV-1a 32bit hash % 100
    public static func hash100(_ key: String) -> Int {
        var hash: UInt32 = 0x811C9DC5
        for b in key.utf8 {
            hash ^= UInt32(b)
            hash = hash &* 0x01000193
        }
        return Int(hash % 100)
    }

    /// percent ∈ [0, 100]；返回 true 代表本 room 命中灰度。
    public static func inGrayPercent(roomID: String, percent: Int) -> Bool {
        if percent <= 0 { return false }
        if percent >= 100 { return true }
        return hash100(roomID) < percent
    }
}
